### PR TITLE
[Mercure] Add the component

### PR DESCRIPTION
**Pull request by dunglas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no     
| Deprecations? | no 
| Tests pass?   | yes    
| Fixed tickets | n/a   
| License       | MIT
| Doc PR        | todo

Last week, I introduced [the Mercure protocol](https://mercure.rocks) as well as [a reference implementation of a mercure hub](https://github.com/dunglas/mercure#hub-implementation).

Mercure allows to push updates from servers to clients, through a hub. It uses [server-sent events](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events), so it passes through all firewalls, works even with very old browsers, is super efficient when using HTTP/2 and is natively supported in all major browsers (no SDK required). [Try the demo](https://demo.mercure.rocks/).

For instance, Mercure allows to automatically and instantly update all currently connected clients (web apps, mobile apps...) every time a resource is modified, created or deleted (e.g. when a `POST` request occurs).

This PR introduce a new component, allowing to use Mercure in Symfony projects in a very convenient way.

# Minimal example

```php
<?php

// dipatch an update from Symfony (for instance in a POST controller)
namespace App\Controller;

use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\Messenger\MessageBusInterface;
use Symfony\Component\Mercure\Update;

class Publish
{
    public function __invoke(MessageBusInterface $bus)
    {
        // use doctrine to save an event
        $bus->dispatch(new Update(
            'http://example.com/foo/bar',
            'New content of the resource'
        ));

        return new Response('OK');
    }
}
```

```javascript
// subscribe to updates client side, can be in a Twig template, in a Progressive Web App...
const eventSource = new EventSource('https://demo.mercure.rocks/subscribe?topic=http://example.com/foo/bar');

// The callback will be called every time an update is published
eventSource.onmessage = e => console.log(e); // do something with the payload, for instance update the view
```

Here we use the Messenger component to dispatch the update. It means that the request to the Mercure hub can be sent synchronously (by default), or asynchronously in a worker if you configure a transport such as RabbitMQ or Reddis.

The corresponding config:

```yaml
# config
framework:
    mercure:
        hubs:
            default:
                url: 'https://demo.mercure.rocks/hub'
                jwt: '%env(MERCURE_JWT)%' # The publisher JWT, provided by the hub
```

# Subscribing to several topics

Mercure allows to subscribe to several topics, and to topics matching a given pattern:

```javascript
// client side
const hub = new URL('https://demo.mercure.rocks/hub');
hub.searchParams.append('topic', 'https://example.com/foo/{id}'); // Templated IRI, any topic matching it will be received
hub.searchParams.append('topic', 'https://example.com/another/resource');

const eventSource = new EventSource(hub);
eventSource.onmessage = e => console.log(e); // do something with the payload
```

# Making the Hub auto-discoverable

Mercure can leverage the web linking RFC to advertise available hubs to the clients:

```php
<?php

namespace App\Controller;

use Fig\Link\Link;
use Lcobucci\JWT\Builder;
use Lcobucci\JWT\Signer\Hmac\Sha256;
use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\HttpFoundation\Request;

class Discover extends AbstractController
{
    public function __invoke(Request $request)
    {
        $this->addLink($request, new Link('mercure', 'https://demo.mercure.rocks'));

        return $this->json(['@id' => 'http://example.com/foo/bar', 'availability' => 'https://schema.org/InStock']);
    }
}
```

In the previous example, the Symfony WebLink Component is used to generate the appropriate `Link` header. The hub can then be discovered using [some lines of JS](https://github.com/dunglas/mercure#examples).

Requires #28875.

# Authorization

Mercure also allows to securely dispatch updates only to users having a specific username, role or group... you name it. To do so, you just have to set a JSON Web Token containing the list of targets in a claim named `mercureTargets`. This token must be stored in a cookie named `mercureAuthorization` (the protocol also allows to transmit the JWT using OAuth, OpenID Connect, and any other transport).

```php
<?php

namespace App\Controller;

use Fig\Link\Link;
use Lcobucci\JWT\Builder;
use Lcobucci\JWT\Signer\Hmac\Sha256;
use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\HttpFoundation\JsonResponse;
use Symfony\Component\HttpFoundation\Request;

class Discover extends AbstractController
{
    public function __invoke(Request $request)
    {
        // ...
        $username = $this->getUser()->getUsername();
        $token = (new Builder())
            // set other JWT appropriate JWT claims, such as an expiration date
            ->set('mercure', ['subscribe' => [$username, 'another-target']]) // could also include the security roles
            ->sign(new Sha256(), '!UnsecureChangeMe!') // store the key in a parameter instead
            ->getToken();

        $response = new JsonResponse(['@id' => '/demo/books/1.jsonld','availability' => 'https://schema.org/InStock']);
        $response->headers->set('set-cookie', sprintf('mercureAuthorization=%s; path=/subscribe; secure; httponly; SameSite=strict', $token));

        return $response;
    }
}
```

```php
<?php

namespace App\Controller;

use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\Messenger\MessageBusInterface;
use Symfony\Component\Mercure\Update;

class Publish
{
    public function __invoke(MessageBusInterface $bus)
    {
        // use doctrine to save an event
        $id = $bus->dispatch(new Update(
            'http://example.com/foo/bar',
            'New content of the resource',
            ['a-username', 'a-group', 'anything-else'] // the targets authorized to receive this update
        ));

        return new Response($id, 200, ['Content-Type' => 'text/plain']);
    }
}
```

That's all! Only users having one of the specified targets will receive the update. In this example, [`lcobucci/jwt`](https://github.com/lcobucci/jwt) is used, but feel free to use any other library to generate the token.

To use the authorization feature, the hub and the Symfony app must be served from the same domain name (can be different subdomains).

### CODE REVIEW

**Comment by javiereguiluz** [comment]
If this component is accepted, I wonder if we should rename it to a more generic name describing its purpose instead of using the name of the provider used. Laravel for example calls this "Broadcasting" and Rails calls it "ActionCable". Thanks!

**Comment by dunglas** [comment]
@javiereguiluz actually, this is not an abstraction layer but an implementation of the protocol (like HttpFoundation implements the HTTP protocol or WebLink implements Web Linking).
It should indeed be possible to create publishers for other transports, but in this case it would be nice to identify which ones, to be sure we can implement them.

**Review by OskarStark — APPROVED** [review]

File: src/Symfony/Component/Mercure/Publisher.php [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+final class Publisher
+{
+    private $publishEndpoint;
+    private $jwtProvider;
+    private $httpClient;
+
+    public function __construct(string $publishEndpoint, callable $jwtProvider, callable $httpClient = null)
```
[the remark this answers is not in the document]
Comment (dkarlovi) [reply src/Symfony/Component/Mercure/Publisher.php]: Possibly in the format used by Psalm and PHPStan IIRC which have the most chance becoming the standard.

File: src/Symfony/Component/Mercure/Publisher.php [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        // All Mercure's keys are safe, so don't need to be encoded, but it's not a generic solution
+        return sprintf('%s=%s', $key, urlencode($value));
+    }
+
+    private function publish(string $url, string $jwt, string $postData)
+    {
+        $result = @file_get_contents($this->publishEndpoint, false, stream_context_create(array('http' => array(
+            'method' => 'POST',
+            'header' => "Content-type: application/x-www-form-urlencoded\r\nAuthorization: Bearer $jwt",
```
[the remark this answers is not in the document]
Comment (dunglas) [reply src/Symfony/Component/Mercure/Publisher.php]: The JWT is now checked using a regex.

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd [hunk 3]
```diff
@@ -417,4 +418,19 @@
         </xsd:sequence>
         <xsd:attribute name="id" type="xsd:string" use="required"/>
     </xsd:complexType>
+
+    <xsd:complexType name="mercure">
+        <xsd:sequence>
+            <xsd:element name="hub" type="mercure_hub" minOccurs="0" maxOccurs="unbounded" />
```
[the remark this answers is not in the document]
Comment (dunglas) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd]: Not in XML:

```xml
<hub name="default" url="https://demo.mercure.rocks/hub" jwt="foo" />
<hub name="default" url="https://anotherhub.com/hub" jwt="bar" />
```

File: src/Symfony/Component/Mercure/Publisher.php [hunk 4]
```diff
@@ -80,16 +87,32 @@ private function encode($key, $value): string
...
         if (false === $result) {
-            throw new \RuntimeException(sprintf('Unable to publish the update to the Mercure hub: %s', error_get_last()));
+            throw new \RuntimeException(sprintf('Unable to publish the update to the Mercure hub: %s', error_get_last()['message'] ?? 'unknown error'));
         }
 
         return $result;
     }
+
+    /**
+     * Regex ported from Windows Azure Active Directory IdentityModel Extensions for .Net.
```
Comment (OskarStark) [inline src/Symfony/Component/Mercure/Publisher.php]: s/.net/.NET/

**Review by OskarStark — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Mercure/Publisher.php (hunk 4 above)
Comment (dunglas) [reply src/Symfony/Component/Mercure/Publisher.php]: I copied the name of the project from their README: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet#windows-azure-active-directory-identitymodel-extensions-for-net

**Review by OskarStark — APPROVED** [review]

**Comment by s7anley** [comment]
I'm just curious why Mercure should be part of Framework bundle? Why not integrate it by own bundle?

**Comment by dunglas** [comment]
@s7anley I've no strong opinion about that. I added it to FrameworkBundle because it's what we do most of the time, but not all times (SecurityBundle, WebServerBundle...). Extracting it in a custom bundle will also allow to extract the library out of `symfony/symfony` at least for one time so I tend to be 👍

**Comment by dunglas** [comment]
The component, and the related bundle, are now available as standalone packages:

* https://github.com/symfony/mercure
* https://github.com/symfony/mercure-bundle

**Pull request closed without merging**